File: 20.Trie/Trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.isEnd = True
        print(f"Inserted: {word}")

    def search(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return node.isEnd

    def _remove_helper(self, node, word, depth):
        if not node:
            return False
        if depth == len(word):
            if not node.isEnd:
                return False
            node.isEnd = False
            return len(node.children) == 0
        ch = word[depth]
        if ch in node.children and self._remove_helper(node.children[ch], word, depth + 1):
            del node.children[ch]
            return not node.isEnd and len(node.children) == 0
        return False

    def remove(self, word):
        if not self.search(word):
            print(f"Not found: {word}")
        else:
            self._remove_helper(self.root, word, 0)
            print(f"Removed: {word}")


trie = Trie()
while True:
    try:
        choice = int(input())
        if choice == 4:
            break
        word = input().strip()
        if choice == 1:
            trie.insert(word)
        elif choice == 2:
            print(("Found: " if trie.search(word) else "Not found: ") + word)
        elif choice == 3:
            trie.remove(word)
    except EOFError:
        break

File: 20.Trie/Trie.java
import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEnd;
}

class Trie {
    TrieNode root = new TrieNode();

    void insert(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray())
            node = node.children.computeIfAbsent(ch, c -> new TrieNode());
        node.isEnd = true;
        System.out.println("Inserted: " + word);
    }

    boolean search(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            node = node.children.get(ch);
            if (node == null) return false;
        }
        return node.isEnd;
    }

    boolean removeHelper(TrieNode node, String word, int depth) {
        if (node == null) return false;
        if (depth == word.length()) {
            if (!node.isEnd) return false;
            node.isEnd = false;
            return node.children.isEmpty();
        }
        char ch = word.charAt(depth);
        TrieNode next = node.children.get(ch);
        if (removeHelper(next, word, depth + 1)) {
            node.children.remove(ch);
            return !node.isEnd && node.children.isEmpty();
        }
        return false;
    }

    void remove(String word) {
        if (!search(word))
            System.out.println("Not found: " + word);
        else {
            removeHelper(root, word, 0);
            System.out.println("Removed: " + word);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Trie trie = new Trie();

        while (sc.hasNextInt()) {
            int choice = sc.nextInt();
            if (choice == 4) break;
            String word = sc.next();
            switch (choice) {
                case 1 -> trie.insert(word);
                case 2 -> System.out.println((trie.search(word) ? "Found: " : "Not found: ") + word);
                case 3 -> trie.remove(word);
            }
        }
        sc.close();
    }
}
